boot/freeldr/freeldr/arch/uefi/uefimem.c

```diff
@@ -256,19 +260,35 @@ UefiMemGetMemoryMap(ULONG *MemoryMapSize)
             }
         }
 
-        UefiSetMemory(FreeldrMem,
-                      MapEntry->PhysicalStart,
-                      MapEntry->NumberOfPages,
-                      MemoryType);
+        /* Sometimes our loader can be loaded into higher memory than we ever allocate */
+        if (MemoryType == LoaderLoadedProgram)
+        {
+            if (((MapEntry->PhysicalStart + (MapEntry->NumberOfPages * PAGE_SIZE)) >> EFI_PAGE_SHIFT) > LoaderPagesSpanned)
+            {
+                /* This value needs to be adjusted if this occurs */
+                LoaderPagesSpanned = ((MapEntry->PhysicalStart + (MapEntry->NumberOfPages * PAGE_SIZE)) >> EFI_PAGE_SHIFT);
+            }
+        }
+
+        /* We really don't want to touch these reserved spots at all */
+        if (MemoryType != LoaderReserve)
+        {
+            UefiSetMemory(FreeldrMem,
+                          MapEntry->PhysicalStart,
+                          MapEntry->NumberOfPages,
+                          MemoryType);
+        }
 
         MapEntry = NEXT_MEMORY_DESCRIPTOR(MapEntry, DescriptorSize);
     }
 
+    /* Reserve the first page, as on some UEFI systems it can be free */
+    UefiSetMemory(FreeldrMem, 0x000000, 1, LoaderFirmwarePermanent);
```

Comment:
I know.... the UEFI version of SetMemory takes the 3rd parameter as a size in number of pages, instead of size in number of bytes, as done in the Setmemory functions of the other "platforms" (BIOS, PC-98, XBOX).
Why this difference? Is it possible to go back to size in bytes (and then the function would divide by the size of one page for the UEFI platform :thinking: )

Reply:
UEFI gives you the size in pages, and this is UefiSetMemory. I just wanted it to be clean instead of adding * PAGE_SIZE to the UefiSetMemory calls in the loader map.